### ISSUE TITLE
ELM-FATES Two Stream Restart Error

### ISSUE THREAD

**Issue by rgknox** [NGEET/fates#1234]
@ckoven detected a problem restarting the model in-between spin-up phases with ELM-FATES. I believe the last phase to complete was an AD spinup, which is usually followed by another pre-industrial run without accerlated decomposition constants. This type of run typically restarts via specifying the finidat file.  @ckoven can you provide more details and perhaps share the script that was used to generate the error.
The model fails during two-stream radiation, with what appears to be issues with uninitialized values.

```
 Warning. small nocomp patch wasnt able to find another patch to fuse with.
270:            0           0  4.157895100850780E-003
 19: forrtl: error (65): floating invalid
 19: Image              PC                Routine            Line        Source             
 19: libpthread-2.31.s  00001479D53EF910  Unknown               Unknown  Unknown
 19: e3sm.exe           000000000422A710  twostreammlpemod_         365  TwoStreamMLPEMod.F90
 19: e3sm.exe           00000000042332F8  twostreammlpemod_         562  TwoStreamMLPEMod.F90
 19: e3sm.exe           000000000421CF76  fatestwostreamuti         353  FatesTwoStreamUtilsMod.F90
 19: e3sm.exe           00000000042066A2  fatesradiationdri         432  FatesRadiationDriveMod.F90
 19: e3sm.exe           0000000000BD93ED  elmfatesinterface        2220  elmfates_interfaceMod.F90
 19: e3sm.exe           00000000009C446F  elm_driver_mp_elm         722  elm_driver.F90
 19: e3sm.exe           000000000096124B  lnd_comp_mct_mp_l         617  lnd_comp_mct.F90
 19: e3sm.exe           00000000004AE943  component_mod_mp_         757  component_mod.F90
 19: e3sm.exe           0000000000469F0A  cime_comp_mod_mp_        2963  cime_comp_mod.F90
 19: e3sm.exe           00000000004971EB  MAIN__                    153  cime_driver.F90
 19: e3sm.exe           0000000000433CFD  Unknown               Unknown  Unknown
 19: libc-2.31.so       00001479D4E3C24D  __libc_start_main     Unknown  Unknown
 19: e3sm.exe           0000000000433C2A  Unknown               Unknown  Unknown
```

```
MOSART decomp info proc =       383 begr =    258526 endr =    259200 numr =       675
270:  Warning. small nocomp patch wasnt able to find another patch to fuse with.
270:            0           0  4.157895100850780E-003
345:  GETRDN
345:    27.2995144373317     
345:   0.823811719313825     
345:   0.000000000000000E+000
345:   1.289252691832906E-008
345:   0.000000000000000E+000  2.328840873721851E-014
345:                      NaN  4.143972961881559E-002
345:   3.510709829585243E-014  1.289246852282203E-008
345:    75.0546825251975     
345:    61.7607829908001     
345:    1.00000000000000     
345:    1.00000000000000     
345:  ENDRUN:
345:  ERROR in TwoStreamMLPEMod.F90 at line 388     
```
When combined with the write statement code:

```
 write(log_unit,*)"GETRDN"
scelg%Kb    = 27.2995144373317
scelb%a  = 0.823811719313825
vai  = 0.000000000000000E+000
scelb%Ad  = 1.289252691832906E-008
scelb%B1 =  0.000000000000000E+000  scelb%B2 = 2.328840873721851E-014
scelb%lambda1_beam =   NaN scelb%lambda2_beam = 4.143972961881559E-002
scelb%lambda1_diff = 3.510709829585243E-014  scelb%lambda2_diff = 1.289246852282203E-008
this%band(ib)%Rbeam_atm =  75.0546825251975
this%band(ib)%Rdiff_atm = 61.7607829908001
exp(-scelg%Kb*vai) = 1
exp(scelb%a*vai) = 1
```

**Comment by rgknox** [comment]
What is standing out to me @ckoven is there is no vegetated area in this element. If this is true, there should be no element, unless it is an "air" element, but that seems unlikely.  The Kb value is not 0.5, which is a nominal value given for the air element.

**Comment by ckoven** [comment]
Thanks @rgknox, I have a minimal run script that reproduces the bug with just one year of runtime, I paste it below. Currently I am using [this tag](https://github.com/ckoven/fates/commit/ec6008e89dff75f48e35e5906e528dfb903a9237), which combines a few different things, I can try something closer to main as well.

```
#!/usr/bin/env bash

SRCDIR=$HOME/E3SM/components/elm/src/
cd ${SRCDIR}
GITHASH1=`git log -n 1 --format=%h`
cd external_models/fates
GITHASH2=`git log -n 1 --format=%h`

STAGE=STEP_ONE
#STAGE=STEP_TWO

if [ "$STAGE" = "STEP_ONE" ]; then
    SETUP_CASE=fates_e3sm_perlmttr_4x5_test_step1
elif [ "$STAGE" = "STEP_TWO" ]; then
    SETUP_CASE=fates_e3sm_perlmttr_4x5_test_step2
fi
    
CASE_NAME=${SETUP_CASE}_${GITHASH1}_${GITHASH2}
basedir=$HOME/E3SM/cime/scripts

cd $basedir
export RES=f45_f45
project=m2467

./create_newcase -case ${CASE_NAME} -res ${RES} -compset IELMFATES -mach pm-cpu -project $project

cd $CASE_NAME

ncgen -o fates_params_default_${GITHASH2}.nc ${SRCDIR}/external_models/fates/parameter_files/fates_params_default.cdl

if [ "$STAGE" = "STEP_ONE"  ]; then

    ./xmlchange RUN_STARTDATE=0001-01-01
    ./xmlchange NTASKS=-3
    ./xmlchange STOP_N=1
    ./xmlchange REST_N=1
    ./xmlchange STOP_OPTION=nyears
    ./xmlchange JOB_QUEUE=debug
    ./xmlchange JOB_WALLCLOCK_TIME=00:30:00
    
    cat > user_nl_elm <<EOF
flandusepftdat = '/global/homes/c/cdkoven/scratch/inputdata/fates_landuse_pft_map_4x5_20240206.nc'
use_fates_luh = .true.
use_fates_nocomp = .true.
use_fates_fixed_biogeog = .true.
fates_paramfile = '${basedir}/${CASE_NAME}/fates_params_default_${GITHASH2}.nc'
use_fates_sp = .false.
fates_spitfire_mode = 1
fates_harvest_mode = 'no_harvest'
use_fates_potentialveg = .true.
fluh_timeseries = ''
EOF

elif [ "$STAGE" = "STEP_TWO" ]; then

    ./xmlchange RUN_STARTDATE=0001-01-01
    ./xmlchange RESUBMIT=0
    ./xmlchange NTASKS=-3
    ./xmlchange STOP_N=1
    ./xmlchange REST_N=1
    ./xmlchange STOP_OPTION=nyears
    ./xmlchange JOB_QUEUE=debug
    ./xmlchange JOB_WALLCLOCK_TIME=00:30:00
    
    cat > user_nl_elm <<EOF
flandusepftdat = '/global/homes/c/cdkoven/scratch/inputdata/fates_landuse_pft_map_4x5_20240206.nc'
use_fates_luh = .true.
use_fates_nocomp = .true.
use_fates_fixed_biogeog = .true.
fates_paramfile = '${basedir}/${CASE_NAME}/fates_params_default_${GITHASH2}.nc'
use_fates_sp = .false.
fates_spitfire_mode = 1
fates_harvest_mode = 'no_harvest'
use_fates_potentialveg = .true.
fluh_timeseries = ''
finidat = '/global/homes/c/cdkoven/scratch/e3sm_scratch/pm-cpu/fates_e3sm_perlmttr_4x5_test_step1_e2fe5d841e_ec6008e8/run/fates_e3sm_perlmttr_4x5_test_step1_e2fe5d841e_ec6008e8.elm.r.0002-01-01-00000.nc'
EOF

fi


./case.setup
./case.build
./case.submit
```

**Comment by glemieux** [comment]
@ckoven I ran the above script on Perlmutter using e3sm commit `bde8cf51ab` and fates tag `sci.1.77.2_api.36.0.0`.  I was able to succesfully run both stages without issue.  For good measure I also adapted the script for ctsm and tested it on derecho without issue as well.  

Per our discussion this morning, would you point me to Rosie's parameter file that you were using so that I can test that next?

**Comment by ckoven** [comment]
@glemieux the updated parameter file is in commit 391d6e9

**Comment by glemieux** [comment]
@ckoven I'm getting a passing case for both stages using the updated parameter file with the same e3sm commit and fates tag as noted above.  

I noticed that the `fluh_timeseries` is being set to `''` in stage 2.  Is that correct for this test?

perlmutter scratch locations: 
```
/global/homes/g/glemieux/scratch/e3sm_scratch/pm-cpu/replicate_RFparam_step1_bde8cf51ab_240366de
/global/homes/g/glemieux/scratch/e3sm_scratch/pm-cpu/replicate_RFparam_step2_bde8cf51ab_240366de
```

**Comment by ckoven** [comment]
Just some updates on this. It is a bit more maddening of a bug than I had originally perceived.  I've now been able to replicate it a few different ways, and it is possibly related to #1237.

The closest-to-main configuration I've gotten is via tag 731bf6ee, which adds three canopy layers and Rosie's parameter file on top of a recent-ish main.  If I run the above script, then the step one part works fine. Step two fails in one of two ways, depending on whether or not `DEBUG=TRUE` in the xml files. If in debug mode, it crashes on reading the restart file with the following:

```
270:  Warning. small nocomp patch wasnt able to find another patch to fuse with.
270:            0           0  4.157895100850780E-003
 30: forrtl: error (65): floating invalid
 30: Image              PC                Routine            Line        Source             
 30: libpthread-2.31.s  0000154235BEF910  Unknown               Unknown  Unknown
 30: e3sm.exe           000000000421D239  twostreammlpemod_         369  TwoStreamMLPEMod.F90
 30: e3sm.exe           0000000004225D4C  twostreammlpemod_         562  TwoStreamMLPEMod.F90
 30: e3sm.exe           000000000420F9C8  fatestwostreamuti         338  FatesTwoStreamUtilsMod.F90
 30: e3sm.exe           00000000041FA476  fatesradiationdri         432  FatesRadiationDriveMod.F90
 30: e3sm.exe           0000000000BD93ED  elmfatesinterface        2220  elmfates_interfaceMod.F90
 30: e3sm.exe           00000000009C446F  elm_driver_mp_elm         722  elm_driver.F90
 30: e3sm.exe           000000000096124B  lnd_comp_mct_mp_l         617  lnd_comp_mct.F90
 30: e3sm.exe           00000000004AE943  component_mod_mp_         757  component_mod.F90
 30: e3sm.exe           0000000000469F0A  cime_comp_mod_mp_        2963  cime_comp_mod.F90
 30: e3sm.exe           00000000004971EB  MAIN__                    153  cime_driver.F90
 30: e3sm.exe           0000000000433CFD  Unknown               Unknown  Unknown
 30: libc-2.31.so       000015423563C24D  __libc_start_main     Unknown  Unknown
 30: e3sm.exe           0000000000433C2A  Unknown               Unknown  Unknown
```

If debug is not true, then it makes it several months into the run, before crashing with a longer error message, the first few lines of which are below. This is the same failure mode that goes away in the longer continuous ruins if I apply the promotion reordering fix in #1237. But the really weird part is that if I apply the promotion reordering fix in #1237 to this script, then the model crashes in the way that  I originally encountered in this bug.
```
141:  Total canopy flux balance not closing in TwoStrteamMLPEMod:Solve
141:  Relative Error, delta/(Rbeam_atm+Rdiff_atm) :  0.895974502301465     
141:  Max Error:   1.000000000000000E-006
141:  ib:            1
141:  scattering coeff:   0.178200000000000     
141:  Breakdown:           2
141:                 1           1
141:        0.164447326258415               12  0.467184519382494     
141:        0.391912784682456       0.929927898941276       0.966422554120326     
```

So TLDR, this does not seem to be related to the grazing code. I will try to go back closer to main on this, first to see if the crash in debug mode happens with just two potential canopy layers but the modified parameter file, and also to see if it happens on main with default parameters.

**Comment by ckoven** [comment]
OK, looked more into the minimum set of things needed to trigger this.  I get the same failure mode (crash during initialization from a prior otherwise-identical one-year-long run's restart file when debug set to true) if `nclmax=2` and using the modified parameter file. It doesn't happen on main with default parameter file.  But if I set to use two stream in the parameter file, then it happens. So somehow that seems to be either causing the crash or exposing an error generated somewhere else.

**Comment by rgknox** [comment]
That nan-check is inside a local debug, which is only true when the debug logical in the file is hard-coded to true:

https://github.com/NGEET/fates/blob/main/radiation/TwoStreamMLPEMod.F90#L36

**Comment by ckoven** [comment]
> I wonder if the simple equivalency test that we commented out will catch it:

Thanks I can try that

**Comment by ckoven** [comment]
OK that worked to catch the error without setting debug in the xml file:

```
206:  GETRDN
206:    27.5512011283433     
206:   0.826620635513254     
206:   0.000000000000000E+000
206:   0.000000000000000E+000
206:   0.000000000000000E+000  0.000000000000000E+000
206:   0.000000000000000E+000  0.000000000000000E+000
206:                      NaN  0.000000000000000E+000
206:    44.7335246941840     
206:    51.3457894877309     
206:    1.00000000000000     
206:    1.00000000000000     
206:  ENDRUN:
206:  ERROR in TwoStreamMLPEMod.F90 at line 388                                      
206:                                                                                 
206:                                                                                 
206:                                                                                 
206:                                                                                 
206:                                                                                 
206:                                        
206:  ERROR: Unknown error submitted to shr_abort_abort.
206: Image              PC                Routine            Line        Source             
206: e3sm.exe           0000000001455CBD  shr_abort_mod_mp_         114  shr_abort_mod.F90
206: e3sm.exe           0000000000FD65B3  twostreammlpemod_         258  TwoStreamMLPEMod.F90
206: e3sm.exe           0000000000FD703F  twostreammlpemod_         562  TwoStreamMLPEMod.F90
206: e3sm.exe           0000000000FD3AA0  fatestwostreamuti         338  FatesTwoStreamUtilsMod.F90
206: e3sm.exe           0000000000FD1064  fatesradiationdri         432  FatesRadiationDriveMod.F90
206: e3sm.exe           00000000005F4950  elmfatesinterface        2220  elmfates_interfaceMod.F90
206: e3sm.exe           00000000005725D1  elm_driver_mp_elm         722  elm_driver.F90
206: e3sm.exe           0000000000558570  lnd_comp_mct_mp_l         617  lnd_comp_mct.F90
206: e3sm.exe           000000000045D42E  component_mod_mp_         757  component_mod.F90
206: e3sm.exe           0000000000436D14  cime_comp_mod_mp_        2963  cime_comp_mod.F90
206: e3sm.exe           000000000045D0C2  MAIN__                    153  cime_driver.F90
206: e3sm.exe           000000000043447D  Unknown               Unknown  Unknown
206: libc-2.31.so       000014D219E3C24D  __libc_start_main     Unknown  Unknown
206: e3sm.exe           00000000004343AA  Unknown               Unknown  Unknown
206: MPICH ERROR [Rank 206] [job id 29651724.0] [Wed Aug 21 15:40:26 2024] [nid006876] - Abort(1001) (rank 206 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 1001) - process 206
206: 
206: aborting job:
206: application called MPI_Abort(MPI_COMM_WORLD, 1001) - process 206
```

**Comment by ckoven** [comment]
Another thing that I see and am not sure how to interpret. I thought that this was happening during the first timestep but it isn't, it's a bit afterwards, as seen in the lnd.log:

``` 
dtime_sync=         1800  dtime_elm=         1800  mod =            0
 Beginning timestep   : 0001-01-01_00:00:00
 --WARNING-- skipping CN balance check for first timestep
 --WARNING-- skipping CN balance check for first timestep
    Completed timestep: 0001-01-01_00:00:00
 Beginning timestep   : 0001-01-01_00:30:00
 --WARNING-- skipping CN balance check for first timestep
 FATES dynamics start
 FATES dynamics complete
 --WARNING-- skipping CN balance check for first timestep
    Completed timestep: 0001-01-01_00:30:00
 Beginning timestep   : 0001-01-01_01:00:00
    Completed timestep: 0001-01-01_01:00:00
<CRASH>
```

**Comment by ckoven** [comment]
to add further confusion, I tried adding a NaN check for `scelb%lambda1_diff`, the variable showing up as a NaN above, in the one place it appears on the LHS via  https://github.com/ckoven/fates/commit/e4612c02b44d36c2e77421d20224522dcf381ad2 but it didn't catch anything.  :(

**Comment by rgknox** [comment]
That scattering element has nonsense values in it anyway, I'm not surprised its generating nonsense results for a scattering solution.   For instance, it has a beam optical depth of exactly zero, which should be impossible if there is material to scatter.

Here is where the calculation is made:
https://github.com/NGEET/fates/blob/main/radiation/TwoStreamMLPEMod.F90#L984

My best guess is  that we have a situation where there are cohorts with no stem or leaf area. No leaf area is normal, ie deciduous trees, but no stem area should be impossible... Unless this uses the new grass allometry that has no SAI or LAI?

I know that line 984 above, does divide by the leaf and stem area, so it should be an inf or a nan, but its possible the compiler is protecting the division because its a 0 divided by a 0.

**Comment by ckoven** [comment]
Thanks @rgknox. So if I work upstream setting traps to try to identify scattering elements where lai and sai are both zero, that might help identify what is going on?

**Comment by rgknox** [comment]
I would put a trap here right after the lai and sai of the element is calculated, if it is less than nearzero, I would fail and print out the cohort involved.

Like here:
https://github.com/NGEET/fates/blob/main/radiation/FatesTwoStreamUtilsMod.F90#L213

**Comment by ckoven** [comment]
Thanks again @rgknox. OK I added some more diagnostics to the error message that it was crashing at before, because I noticed that if VAI + LAI was near zero, then logic just above the line you point to should have already made the pft of the scattering element the `air_ft`.  So anyway, seems even weirder. 

```
206:  GETRDN
206:    27.5512011283433     
206:   0.826620635513254     
206:   0.000000000000000E+000
206:   0.000000000000000E+000
206:   0.000000000000000E+000  0.000000000000000E+000
206:   0.000000000000000E+000  0.000000000000000E+000
206:                      NaN  0.000000000000000E+000
206:    44.7335246941840     
206:    51.3457894877309     
206:    1.00000000000000     
206:    1.00000000000000     
206:  ican           1
206:  icol          10
206:  ib           1
206:  scelg%pft           5
206:  scelg%lai   1.99931012632750     
206:  scelg%sai  0.184394986794637     
206:  ENDRUN:
206:  ERROR in TwoStreamMLPEMod.F90 at line 394                                      
206:                                                                                 
206:                                                                                 
206:                                                                                 
206:                                                                                 
206:                                                                                 
206:                                        
206:  ERROR: Unknown error submitted to shr_abort_abort.
206: Image              PC                Routine            Line        Source             
206: e3sm.exe           000000000145613D  shr_abort_mod_mp_         114  shr_abort_mod.F90
206: e3sm.exe           0000000000FD6802  twostreammlpemod_         258  TwoStreamMLPEMod.F90
206: e3sm.exe           0000000000FD728F  twostreammlpemod_         568  TwoStreamMLPEMod.F90
206: e3sm.exe           0000000000FD3AA0  fatestwostreamuti         338  FatesTwoStreamUtilsMod.F90
206: e3sm.exe           0000000000FD1064  fatesradiationdri         432  FatesRadiationDriveMod.F90
206: e3sm.exe           00000000005F4950  elmfatesinterface        2220  elmfates_interfaceMod.F90
206: e3sm.exe           00000000005725D1  elm_driver_mp_elm         722  elm_driver.F90
206: e3sm.exe           0000000000558570  lnd_comp_mct_mp_l         617  lnd_comp_mct.F90
206: e3sm.exe           000000000045D42E  component_mod_mp_         757  component_mod.F90
206: e3sm.exe           0000000000436D14  cime_comp_mod_mp_        2963  cime_comp_mod.F90
206: e3sm.exe           000000000045D0C2  MAIN__                    153  cime_driver.F90
206: e3sm.exe           000000000043447D  Unknown               Unknown  Unknown
206: libc-2.31.so       000014E2CB83C24D  __libc_start_main     Unknown  Unknown
206: e3sm.exe           00000000004343AA  Unknown               Unknown  Unknown
```

**Comment by rgknox** [comment]
The vai that it is reporting on the third line, is the integrated vegetation depth (top down coordinate) at which it is being asked to report downwelling radiation in this element. So, its reporting downwelling radiation at the top of the element. The element itself, does have non-zero lai and sai, so that is good, and may help clear up some confusion. Ill look at this more and see if I can figure anything out.

**Comment by rgknox** [comment]
Maybe the sun is at a very low inclination angle, which is why the Kb term is so high. It could be that this is making life difficult for the solver?  

For instance we have exponentials that have Kb * LAI.  This could create math operations of e^(+-100) or greater.  Maybe we just need cap Kb...  An exponential function of even e^(-300) should not generate a number larger than can be held in a 16 byte real, but it could still be making life tough on the matrix inversion.

https://github.com/NGEET/fates/blob/main/radiation/TwoStreamMLPEMod.F90#L506

Try reducing kb_max to something like 10?  
https://github.com/NGEET/fates/blob/main/radiation/TwoStreamMLPEMod.F90#L74



But I think it would be good to also get more info on why some of these other terms are zero... For instance, lets look at everything that builds the "Ad" term...

https://github.com/NGEET/fates/blob/main/radiation/TwoStreamMLPEMod.F90#L1245C11-L1266C82

Can you add print statements to your current fail statement that includes more of the parameters in the line above:

 write(log_unit,*) "om:" ,scelbp%om
write(log_unit,*) "betab: ",scelbp%betab
write(log_unit,*) "Kd: ",scelgp%Kd
write(log_unit,*) "betad: ",scelbp%betad
write(log_unit,*)  "b2 =", -(scelgp%Kd*(1._r8-scelbp%om)*(1._r8-2._r8*scelbp%betab)+scelgp%Kb) * &
               scelbp%om*scelgp%Kb*scelbp%Rbeam0
          
 write(log_unit,*)  "b1 =",  -(scelgp%Kd*(1._r8-scelbp%om+2._r8*scelbp%om*scelbp%betad) + &
               (1._r8-2._r8*scelbp%betab)*scelgp%Kb) * &
               scelbp%om*scelgp%Kb*scelbp%Rbeam0

**Comment by ckoven** [comment]
Thanks @rgknox! the tag that has all these diagnostics is dc16e85e.

**Comment by rgknox** [comment]
In the spirit of this very special issue number, I'd like to remind everyone of this scene from spaceballs: 

https://www.youtube.com/watch?v=a6iW-8xPw3k

This is also a good reminder about password quality while we are at it.

**Comment by rgknox** [comment]
So I dumped all the cohorts when the error is generated (in the GetRdDn() function), and found two things that I thought were quite strange...

1) This is the fifth out of 5 cohorts.  According to the two-stream, there is only 1 canopy layer, and the fractional area of those cohorts sums to exactly 1.  This should be super incredibly massively unlikely.  The area of an upper canopy layer should sum to exactly 1, but there would be a second or more layers below it.  Or, if there is really only 1 layer, it is unlikely that the cohorts "exactly" fill it up, and one element should be an air element...  But why? @ckoven is this possibly a new feature of the potential veg? Could there be exactly 1 canopy layer, perfectly full?  If not, there must be some issue where we are incorrectly dropping the cohorts in the next layer from the scattering element construct. I'll run  a test where I dump out the patch info as well.

2) The Rbeam0 value is 1.0 in the first 4 cohorts, but 0 in the fifth and problematic cohort. The scattering coefficients are almost the same for all five of them.  This is the unit downwelling beam radiation at the top of the cohort. For a single canopy layer, this is very trivial, it should be 1.0.  This value is calculated early on in %Solve(), the core two-stream routine.  So this tells me that for some reason, the 5th cohort simply did not run %Solve() prior to this call. If the scattering coefficients look normalish, I suppose everything but the solve was called...

**Comment by ckoven** [comment]
@rgknox that is very weird that there is only one exactly full canopy layer. I don't see how that could happen unless all understory cohorts are getting killed somehow.  I don't know enough about the call sequence to understand how point 2 could have happened.

**Comment by ckoven** [comment]
There is some discussion right now on the CLM call about the distinction between hybrid runs (equivalent to setting finidat to a prior run's restart file), and branch runs (equivalent to continuing a restart file). Given that this crash happens only in the hybrid type case, it might be helpful to know exactly what happens differently during the hybrid case restart read versus a branch  case restart read.

**Comment by ckoven** [comment]
Just as a sanity check, I ran a slightly modified script from the above in both a `RUN_TYPE=hybrid` and a `RUN_TYPE=branch` case, and the former case crashed while the latter case did not. I hadn't actually run a branch test, only a restart within a case, so wanted to confirm that doing so would in fact pass, which it did.  So clearly there is some logic specific to the type of initialization from restart files that is generating this.

**Comment by rgknox** [comment]
@ckoven , does this congifuration modify, add or remove any patches or cohorts during the restart procedure, aside from reading in the patch and cohort information from the restart file?

If so, we need to be careful to call update_3dpatch_radiation() after everything is settled:

https://github.com/E3SM-Project/E3SM/blob/master/components/elm/src/main/elmfates_interfaceMod.F90#L1926-L1928

That call to update_3dpatch_radiation() is where the scattering element info is prepared, and the solver is run for the first time.  If we perform this action and then modify the patch structure in anyway before exiting the initialization, it will be problematic.

**Comment by ckoven** [comment]
@rgknox I don't believe there is anything in the restart sequence that is different in this configuration than in the standard full-FATES configuration.

**Comment by rgknox** [comment]
It looks like in a hybrid run, the following initialization sequence, which is maybe meant for a cold-start is run:

https://github.com/NGEET/fates/blob/main/main/EDInitMod.F90#L469

Is it possible that something was run twice, or inadvertantly, and that is creating problems during the hybrid run?

**Comment by ckoven** [comment]
Thanks @rgknox does [this line here ](https://github.com/NGEET/fates/blob/main/main/EDInitMod.F90#L413) evaluate as `.true.` in a hybrid run? I don't think we should be entering the code block under that case.

**Comment by rgknox** [comment]
I believe it would indeed evaluate as true, based on this: 

https://github.com/E3SM-Project/E3SM/blob/master/components/elm/src/main/elmfates_interfaceMod.F90#L482-L487

**Comment by ckoven** [comment]
@rgknox that section of code in `set_site_properties` isn't being triggered before the crash, I guess because the call sequence to it is within this block, which excludes hybrid and finidat='filename' cases: https://github.com/E3SM-Project/E3SM/blob/master/components/elm/src/main/elm_initializeMod.F90#L1023-L1037

**Comment by rgknox** [comment]
The finidat will not be ' ', so this block of code will not be triggered for either branch or hybrid.  I'm just guessing, but my expectation was that we would not want cold-start type initializations to happen in this situation, that initializations would happen during the restart() call in elmfates_interfacemod.F90.

On a side-note, I don't like that string compare on the finidat == ' '. It seems awfully vulnerable that an unspecified finidat has to be empty with exactly 1 space?  I feel like we should to a trim and a string length count of one or less.

**Comment by ckoven** [comment]
Just wanted to note here that @rgknox's fixes on the two branches below appear to fix the issue.

https://github.com/rgknox/fates/tree/twostream-restart-bugfixes
https://github.com/rgknox/E3SM/tree/twostr-rest-bugfix

**Comment by rgknox** [comment]
Yes, I don't believe this is an open issue anymore, we can re-open if the problem resurfaces
